lib/Target/LLVM/LLVMToDynamicLib.cpp

```diff
@@ -0,0 +1,303 @@
+    llvm::errs() << "  " << Target.getName() << "\n";
+  }
+
+  // Set target triple if not already set
+  if (module.getTargetTriple().empty()) {
+    std::string defaultTriple = llvm::sys::getDefaultTargetTriple();
+    llvm::outs() << "Setting default target triple: " << defaultTriple << "\n";
+    module.setTargetTriple(defaultTriple);
+  }
+
+  // auto relocModel = llvm::Optional<llvm::Reloc::Model>();
+
+  // Look up the target
+  llvm::outs() << "Target triple for this module:" << module.getTargetTriple()
+               << "\n";
+  llvm::SmallVector<std::string, 0> attrs; // Empty feature list
+
+  // llvm::TargetMachine *targetMachine = llvm::EngineBuilder().selectTarget(
+  // llvm::Triple(module.getTargetTriple()), "x86-64", "generic", attrs);
+  auto targetMachine = createTargetMachine(module.getTargetTriple());
+  if (!targetMachine) {
+    llvm::errs() << "Failed to create TargetMachine for triple: "
+                 << module.getTargetTriple() << "\n";
+    return llvm::failure();
+  }
+
+  // Set data layout
+  module.setDataLayout(targetMachine->createDataLayout());
+
+  // debug info:
+  llvm::outs() << "TargetMachine Info: \n";
+  llvm::outs() << "Triple: " << targetMachine->getTargetTriple().str() << "\n";
+  llvm::outs() << "DataLayout: "
+               << module.getDataLayout().getStringRepresentation() << "\n";
+
+  // Create an output file stream to write the object file
+  std::error_code EC;
+  llvm::raw_fd_ostream out(outputFilename, EC, llvm::sys::fs::OF_None);
+  if (EC) {
+    llvm::errs() << "Error opening output file: " << EC.message() << "\n";
+    return llvm::failure();
+  }
+
+  // Emit object code to the file
+  llvm::legacy::PassManager passManager;
+  passManager.add(
+      new llvm::TargetLibraryInfoWrapperPass(targetMachine->getTargetTriple()));
+  if (targetMachine->addPassesToEmitFile(passManager, out, nullptr,
+                                         llvm::CodeGenFileType::ObjectFile)) {
+    llvm::errs() << "Target machine cannot emit object file\n";
+    return llvm::failure();
+  }
+
+  passManager.run(module);
+
+  return llvm::success();
+}
+
+llvm::LogicalResult runLinkCommand(std::string commandLine) {
+  llvm::dbgs() << "Running linker command:\n" << commandLine << "\n";
+  const auto exitCode = system(commandLine.c_str());
+  if (exitCode == 0) {
+    return llvm::success();
+  }
+  llvm::errs() << "Linking failed; escaped command line returned exit code "
+               << exitCode << ":\n\n"
+               << commandLine << "\n\n";
+  return llvm::failure();
+}
+
+// TODO: decide if we have any use cases where we actually need multiple .o
+// files, or we should just pass in a single 1
+llvm::LogicalResult linkDynamicLibrary(const std::string &libraryName,
+                                       ArrayRef<std::string> objectFileNames,
+                                       bool removeDebugSymbols) {
+  SmallVector<std::string, 8> flags = {
+      "ld.lld-17",
```

Comment:
So I copied some comments from the IREE code below, starting at line 177, but the gist is:
```  
  // While that much is said in the GNU ld man page, the reality is that
  // out of ld.bfd, ld.gold and ld.lld, only ld.lld actually implements
  // that. Meanwhile, ld.bfd interprets -static -shared as just -static,
  // and ld.gold rejects -static -shared outright as "incompatible".
  ```
  I confess I haven't actually tested this for myself, but I thought it seemed reasonable to assume that's accurate and hardcode to lld (+ sort of seems more graceful to stick to LLVM arch all the way down lol).  But I can investigate more if you think it's worthwhile

Reply:
In our case, I wouldn't be surprised if we could omit the `-static` flag since I wouldn't think we would have significant dylib deps of our own, but I could definitely be totally wrong

Reply:
no:
```
ld.bfd     ld.gold    ld.lld-17
```